src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.QueueItem.cs

```diff
@@ -5,59 +5,95 @@
-            /// tracing/logging subsystem.
+            /// Sets the result of this queue item to an exception.  Used if we fail to enqueue this item before it starts.
             /// </summary>
-            public readonly Guid ActivityId;
-            private readonly ILspLogger _logger;
+            public bool TrySetException(Exception e);
 
-            /// <inheritdoc cref="IRequestHandler.MutatesSolutionState" />
-            public readonly bool MutatesSolutionState;
+            /// <summary>
+            /// Cancels this queue item if possible.  Used when the queue is shutting down and needs to be cleared.
+            /// </summary>
+            public bool TrySetCanceled(CancellationToken cancellationToken);
 
             /// <inheritdoc cref="IRequestHandler.RequiresLSPSolution" />
-            public readonly bool RequiresLSPSolution;
+            public bool RequiresLSPSolution { get; }
+
+            /// <inheritdoc cref="IRequestHandler.MutatesSolutionState" />
+            public bool MutatesSolutionState { get; }
 
             /// <inheritdoc cref="RequestContext.ClientName" />
-            public readonly string? ClientName;
-            public readonly string MethodName;
+            public string? ClientName { get; }
 
-            /// <inheritdoc cref="RequestContext.ClientCapabilities" />
-            public readonly ClientCapabilities ClientCapabilities;
+            public string MethodName { get; }
 
             /// <summary>
             /// The document identifier that will be used to find the solution and document for this request. This comes from the <see cref="TextDocumentIdentifier"/> returned from the handler itself via a call to <see cref="IRequestHandler{RequestType, ResponseType}.GetTextDocumentIdentifier(RequestType)"/>.
             /// </summary>
-            public readonly TextDocumentIdentifier? TextDocument;
+            public TextDocumentIdentifier? TextDocument { get; }
+
+            /// <inheritdoc cref="RequestContext.ClientCapabilities" />
+            public ClientCapabilities ClientCapabilities { get; }
 
             /// <summary>
-            /// A cancellation token that will cancel the handing of this request. The request could also be cancelled by the queue shutting down.
+            /// A cancellation token that represents the client's cancellation token for the request.
             /// </summary>
-            public readonly CancellationToken CancellationToken;
+            public CancellationToken RequestCancellationToken { get; }
+
+            /// <summary>
+            /// <see cref="CorrelationManager.ActivityId"/> used to properly correlate this work with the loghub
+            /// tracing/logging subsystem.
+            /// </summary>
+            public Guid ActivityId { get; }
+
+            public RequestMetrics Metrics { get; }
+        }
+
+        private readonly struct QueueItem<TRequestType, TResponseType> : QueueItem
```

Comment:
being a struct isn't really helpful here if it's going to always be used through the itnerface.